File: src/Utility/Pathfinding.js
export default class Pathfinding {
    constructor(terrain, costs) {
        this.terrain = terrain; // Terrain should provide access to the game grid and movement costs
		this.costs = costs
    }

    aStar(start, goal) {
		let openSet = new PriorityQueue((a, b) => a.f < b.f, this._isEqual.bind(this));
		let cameFrom = new Map();
		let gScore = new Map();
		let fScore = new Map();
		let closedSet = new Set();
		let closestNode = start; // Keep track of the node closest to the goal
		let closestNodeDistance = this._heuristic(start, goal);
		
		gScore.set(this._hash(start), 0);
		fScore.set(this._hash(start), closestNodeDistance);
	
		openSet.enqueue({ position: start, f: closestNodeDistance });
		
		while (!openSet.isEmpty()) {
			let current = openSet.dequeue().position;
			// Add current node to closed set
			closedSet.add(this._hash(current));
			if (this._isEqual(current, goal)) {
				return this._reconstructPath(cameFrom, current);
			}
	
			let currentDistance = this._heuristic(current, goal);
			if (currentDistance < closestNodeDistance) {
				closestNode = current;
				closestNodeDistance = currentDistance;
			}
	
			for (let neighbor of this._getNeighbors(current)) {
				// Skip if neighbor is in closed set
				if (closedSet.has(this._hash(neighbor))) {
					continue;
				}
				let tentativeGScore = gScore.get(this._hash(current)) + this._cost(neighbor);

				if (tentativeGScore < (gScore.get(this._hash(neighbor)) || Infinity)) {
					cameFrom.set(this._hash(neighbor), current);
					gScore.set(this._hash(neighbor), tentativeGScore);
					fScore.set(this._hash(neighbor), tentativeGScore + this._heuristic(neighbor, goal));
	
					if (!openSet.contains(neighbor)) {
						openSet.enqueue({ position: neighbor, f: fScore.get(this._hash(neighbor)) });
					}
				}
			}
		}
	
		// If no path is found, check if the closest node is the start node
		if (this._isEqual(closestNode, start)) {
			return []; // Return an empty path if the closest node is the start node
		} else {
			return this._reconstructPath(cameFrom, closestNode); // Return the path to the closest node
		}
	}
	

    _heuristic(nodeA, nodeB) {
        // Use Manhattan distance as the heuristic
        return Math.abs(nodeA.x - nodeB.x) + Math.abs(nodeA.y - nodeB.y);
    }

    _cost(node) {
        // Use the node's position to get the terrain type, and then the cost for that type
        const terrainType = this.terrain[node.y][node.x];
        return this.costs[terrainType];
    }

	_getNeighbors(node) {
        const neighbors = [];
        const directions = [
            { x: 0, y: -1 }, // North
            { x: 1, y: 0 },  // East
            { x: 0, y: 1 },  // South
            { x: -1, y: 0 }  // West
            // Add more directions here if you allow diagonal movement
        ];

        directions.forEach(dir => {
            const neighborX = node.x + dir.x;
            const neighborY = node.y + dir.y;

            // Check if neighbor is within the bounds of the terrain
            if (neighborX >= 0 && neighborX < this.terrain[0].length && neighborY >= 0 && neighborY < this.terrain.length) {
                const terrainType = this.terrain[neighborY][neighborX];
				// Check if the terrain type is passable (cost > 0)
				if (this.costs[terrainType] > 0) {
					neighbors.push({ x: neighborX, y: neighborY });
				}
            }
        });

        return neighbors;
    }

	_reconstructPath(cameFrom, current) {
		let totalPath = [current];
	
		while (cameFrom.has(this._hash(current))) {
			current = cameFrom.get(this._hash(current));
			totalPath.unshift(current);
		}
	
		return totalPath;
	}

    _isEqual(nodeA, nodeB) {
        return nodeA.x === nodeB.x && nodeA.y === nodeB.y;
    }

    _hash(node) {
        return `${node.x},${node.y}`;
    }
}

class PriorityQueue {
    constructor(comparator = (a, b) => a > b, equals = (a, b) => a === b) {
        this._heap = [];
        this._comparator = comparator;
        this._equals = equals; // New equality function		
    }

    enqueue(value) {
        this._heap.push(value);
        this._siftUp();
    }

    dequeue() {
        const result = this._heap[0];
        const end = this._heap.pop();

        if (this._heap.length > 0) {
            this._heap[0] = end;
            this._siftDown();
        }

        return result;
    }

    isEmpty() {
        return this._heap.length === 0;
    }

    contains(node) {
        return this._heap.some(element => this._equals(element.position, node));
    }

    _siftUp() {
        let nodeIndex = this._heap.length - 1;

        while (nodeIndex > 0 && this._compare(nodeIndex, this._parent(nodeIndex))) {
            this._swap(nodeIndex, this._parent(nodeIndex));
            nodeIndex = this._parent(nodeIndex);
        }
    }

    _siftDown() {
        let nodeIndex = 0;

        while (
            (this._left(nodeIndex) < this._heap.length && this._compare(this._left(nodeIndex), nodeIndex)) ||
            (this._right(nodeIndex) < this._heap.length && this._compare(this._right(nodeIndex), nodeIndex))
        ) {
            let maxChild = (this._right(nodeIndex) < this._heap.length && this._compare(this._right(nodeIndex), this._left(nodeIndex))) ? this._right(nodeIndex) : this._left(nodeIndex);

            this._swap(nodeIndex, maxChild);
            nodeIndex = maxChild;
        }
    }

    _parent(index) {
        return ((index - 1) >> 1);
    }

    _left(index) {
        return (2 * index + 1);
    }

    _right(index) {
        return (2 * index + 2);
    }

    _compare(i, j) {
        return this._comparator(this._heap[i], this._heap[j]);
    }

    _swap(i, j) {
        [this._heap[i], this._heap[j]] = [this._heap[j], this._heap[i]];
    }
}

File: server/utils/Pathfinding.py
import heapq
class Pathfinding:
    def __init__(self, terrain, costs):
        self.terrain = terrain  # 2D grid representing the terrain
        self.costs = costs      # Dictionary mapping terrain types to movement costs

    def a_star(self, start, goal):
        goal = self.modify_goal_if_unreachable(start, goal)
        if start == goal:
            #print('start is goal', start, goal)
            return [start]  # or [goal], since they are the same

        # If the goal is an immediate neighbor of the start and is accessible
        if goal in self.get_neighbors(start):
            #print('immediate neighbor', start, goal)
            return [start, goal]
        frontier = PriorityQueue()
        frontier.put(start, 0)
        came_from = {start: None}
        cost_so_far = {start: 0}

        while not frontier.empty():
            current = frontier.get()

            if current == goal:
                break

            for next in self.get_neighbors(current):
                new_cost = cost_so_far[current] + self.cost(next)
                if next not in cost_so_far or new_cost < cost_so_far[next]:
                    cost_so_far[next] = new_cost
                    priority = new_cost + self.heuristic(goal, next)
                    frontier.put(next, priority)
                    came_from[next] = current

        return self.reconstruct_path(came_from, start, goal)

    def heuristic(self, nodeA, nodeB):
        # Use Manhattan distance as the heuristic
        return abs(nodeA[0] - nodeB[0]) + abs(nodeA[1] - nodeB[1])

    def cost(self, node):
        # Use the node's position to get the terrain type, and then the cost for that type
        terrainType = self.terrain[node[1]][node[0]]
        return self.costs[terrainType]

    def get_neighbors(self, node):
        neighbors = []
        directions = [(0, -1), (0, 1), (-1, 0), (1, 0)]
        for dx, dy in directions:
            newX, newY = node[0] + dx, node[1] + dy
            if 0 <= newX < len(self.terrain[0]) and 0 <= newY < len(self.terrain):
                if self.costs[self.terrain[newY][newX]] > 0:  # Check if passable
                    neighbors.append((newX, newY))
        return neighbors

    def reconstruct_path(self, came_from, start, goal):
        if goal not in came_from:
        # Handle the case where no path was found
            return []  # or some other indication that no path was found

        current = goal
        path = []
        while current != start:
            path.append(current)
            current = came_from[current]
        path.append(start)
        path.reverse()
        return path

    def modify_goal_if_unreachable(self, start, goal):
        if not self.is_passable(goal):
            return self.find_nearest_accessible_tile(goal, start)
        return goal

    def is_passable(self, node):
        return self.costs[self.terrain[node[1]][node[0]]] > 0

    def find_nearest_accessible_tile(self, target, start):
        neighbors = self.get_neighbors(target)
        # Sort neighbors by their distance to the start position
        neighbors.sort(key=lambda n: self.heuristic(n, start))
        for neighbor in neighbors:
            if self.is_passable(neighbor):
                return neighbor
        return start  # Fallback in case no accessible neighbors are found

class PriorityQueue:
    def __init__(self):
        self.elements = []
    
    def empty(self):
        return len(self.elements) == 0

    def put(self, item, priority):
        heapq.heappush(self.elements, (priority, item))

    def get(self):
        return heapq.heappop(self.elements)[1]
